.github/workflows/release.yml

```diff
@@ -50,3 +54,59 @@ jobs:
           files: |
             pie.phar
             pie.phar.asc
+
+  docker-binary-only-image:
+    needs: release-phar
+    name: Docker binary-only image
+    runs-on: ubuntu-latest
+    if: ${{ startsWith(github.ref, 'refs/tags/') }}
+
+    permissions:
+      contents: read
+      packages: write
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Log in to the Container registry
+        uses: docker/login-action@v3
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Make image lowercase
+        run: |
+          echo IMAGE_NAME_LOWER=$(echo ${{ env.IMAGE_NAME }} | tr '[:upper:]' '[:lower:]') >> ${GITHUB_ENV}
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          flavor: |
+            latest=true
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME_LOWER }}
+          tags: |
+            type=semver,pattern={{version}}
+            type=semver,pattern={{major}}.{{minor}}
+            type=semver,pattern={{major}}
```

Comment:
This is standard practice for `php` images too, you can pull `8.4.1`, `8.4` or just `8`. Composer and PHP-CS-Fixer also have this. I don't agree with your comment, this works as expected and offers 3 levels of stability, so users can choose how they want install the tool.

Reply:
I think this is fine as is, if people want to use 1.2.3, or latest, or 1, or 1.2, they can. It is a reasonable trade-off IMO. With a tool like this, and Composer, *typically* (if you're working on modern applications), the latest is fine, but this allows people to pin to specific dependencies. Issues come with re-tagging (which I would strongly protest against, causes more issues than good), or retroactively tagging earlier releases, but this is very edge case.

Reply:
> Issues come with re-tagging (which I would strongly protest against, causes more issues than good), or retroactively tagging earlier releases, but this is very edge case.

This 💯. With proper release cycle it just works. Even such edge case can be covered by running the workflow from the Github UI with explicit params, but I believe it won't be needed.

Reply:
> the latest is fine, but this allows people to pin to specific dependencies

Yes. But I'm arguing to think hard about whether we want to encourage users to pin to PIE 1.2.x instead of any PIE 1.x.

For PHP itself it's exactly opposite: I consider the availability of `php:8` to be a mistake. PHP does not follow Semver and generally upgrading from one PHP branch to the next is not trivial, because there's always *some* breaking changes. So users should be encouraged to make an explicit choice when changing the release branch, allowing them to use `php:8` encourages the Semver myth.

Or as another example: HAProxy does not follow Semver either and instead simply treats version numbers as decimals. That means after x.9 comes (x+1).0. There was a `haproxy:1` under the false assumption that any HAProxy 2 would be a major release. With HAProxy 2.0 the “major version” variant was removed. There's only `haproxy:2.0` and `haproxy:2.0.0`. And nowadays also `haproxy:lts` to pin to the latest LTS version.

----------------

Any choice we make at this point will need to be maintained for an indefinite future, because once the versioning scheme cat is out of the bag it's hard to reverse it. By being conservative for now, we can first observe how using the Docker image works for for the users.

Reply:
PHP or HAProxy versioning aside (out of the scope), what's the actual problem with availability of fine-tuned tags for PIE's Docker images? Is it against the versioning policy, in this case SemVer, that I believe is being used in this project? My perspective is:

- `pie:X-bin` is for people who always want to use latest semver-compatible version
- `pie:X.Y-bin` is for people who does not want to deprecation warnings pop up randomly during build (deprecations are supposed to be in minor versions)
- `pie:X.Y.Z-bin` is for people who want to pin-point exact version to have stable and reproducible builds

I believe all those strategies are valid, and having all the tags make it possible to choose the most fitting one.

`latest-bin` (or just `bin`, even though I am not in favour of this) on the other hand is for YDD (YOLO Driven Development) or experiments, it's also good to have it, since it does not cost anything 🙂.